crates/sui-graphql-rpc/src/types/dot_move/named_type.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+
+use async_graphql::Context;
+use futures::future;
+use move_core_types::parser::parse_type_tag;
+use sui_types::base_types::ObjectID;
+
+use crate::error::Error;
+
+use super::{
+    config::{DotMoveServiceError, VERSIONED_NAME_UNBOUND_REG},
+    named_move_package::NamedMovePackage,
+};
+
+pub(crate) struct NamedType;
+
+impl NamedType {
+    pub(crate) async fn query(
+        ctx: &Context<'_>,
+        name: &str,
+        checkpoint_viewed_at: u64,
+    ) -> Result<String, Error> {
+        // we do not de-duplicate the names here, as the dataloader will do this for us.
+        let names = Self::parse_names(name)?;
+
+        // Gather all the requests to resolve the names.
+        let names_to_resolve = names
+            .iter()
+            .map(|x| NamedMovePackage::query(ctx, x, checkpoint_viewed_at))
+            .collect::<Vec<_>>();
+
+        // now we resolve all the names in parallel (data-loader will do the proper de-duplication / batching for us)
+        // also the `NamedMovePackage` query will re-validate the names (including max length, which is not checked on the regex).
+        let mut results = future::try_join_all(names_to_resolve).await?;
+
+        // now let's create a hashmap with {name: MovePackage}
+        let mut name_package_id_mapping = HashMap::new();
+
+        // doing it in reverse so we can pop instead of shift
+        for name in names.into_iter().rev() {
+            // safe unwrap: we know that the amount of results has to equal the amount of names.
```

Comment:
```suggestion
            // SAFETY: we know that the amount of results has to equal the amount of names.
```

Convention is to use `SAFETY` in all caps, so it stands out. Some editors will also highlight this.

Reply:
But the best way to do this is to use `zip`:

```rust
for (name, package) in names.into_iter().zip(results.into_iter()) {
    /* ... */
}
```

And even better, you can then ditch the loop entirely:

```rust
let name_package_id_mapping = HashMap::from_iter(
    names.into_iter().zip(results.into_iter())
);
```

And best is to avoid this problem altogether by introducing a `NamedMovePackage::multi_query` that uses `DataLoader::load_many`.